### DISCUSSION TITLE
How to use the windows crate MEMORY_BASIC_INFORMATION

### DISCUSSION THREAD

**Discussion by TallonIV** [fzyzcjy/flutter_rust_bridge#1787] [Q&A]
Hello, I am asking for help to generate the bindings for flutter rust bridge in my code that uses the windows crate _https://crates.io/crates/windows_, I was advised to wrap it using a mutex. I tried to implement it:

```
pub use std::sync::Mutex;
pub use windows::Win32::System::Memory::MEMORY_BASIC_INFORMATION;
use windows::Win32::System::Memory::VirtualQueryEx;

pub fn memory_regions() -> io::Result<Mutex<MEMORY_BASIC_INFORMATION>>{
    let mut info = std::mem::MaybeUninit::uninit();
    let handle =open_handle(4496);
    let written = unsafe {
        VirtualQueryEx(
            handle,
            Some(std::ptr::null()),
            info.as_mut_ptr(),
            std::mem::size_of::<MEMORY_BASIC_INFORMATION>(),
        )
    };
    if written == 0{
        return Err(std::io::Error::last_os_error())
    }
    close_handle(handle);
    unsafe { Ok(Mutex::new(info.assume_init())) }
}
```
But, when I compile it, I still get the same error:
```
389  | / flutter_rust_bridge::frb_generated_moi_arc_impl_value!(
390  | |     flutter_rust_bridge::for_generated::rust_async::RwLock<Mutex<MEMORY_BASIC_INFORMATION>>
391  | | );
     | |_^ `*mut c_void` cannot be sent between threads safely
```
and `Trait `SseEncode` is not implemented for `Error` [E0277]`
I investigated and I found that it is needed to type `unsafe impl Send for MEMORY_BASIC_INFORMATION`,
but I get another error that I can´t do that because it is an external crate.
How could I fix it?

**Comment by fzyzcjy** [comment]
> `*mut c_void` cannot be sent between threads safely

Hmm, so seems that `MEMORY_BASIC_INFORMATION` is even not Send.

One way is to add some unsafe wrapper, e.g. `unsafe impl Send`, to make it work; or, more safely, ask the authors of windows crate whether it is OK to add `impl Send`. It is because, if a struct is not Send, then it must be used within one single thread from the beginning to the end, which is not possible in Dart (its main thread even changes as time goes by).

There is another way. You may create a separate thread, and do all your work on that single thread from beginning to the end. This approach is a bit like the classical event loop things (e.g. you know, the android main thread even loop, the ios one, the flutter engine one, ...).

> Trait SseEncodeis not implemented for Error

Maybe try to use `anyhow::Result` (or if you have reasons to use io::Result feel free to discuss!)

For future readers: Context - https://github.com/fzyzcjy/flutter_rust_bridge/issues/1784

**Comment by TallonIV** [comment]
Thanks for the answer, I am going to try to investigate more.

**Reply by fzyzcjy** [reply]
You are welcome!
